src/aspire/abinitio/commonline_c3_c4.py

```diff
@@ -231,21 +237,270 @@ def _estimate_inplane_rotations(
     # Secondary Methods for computing outer product #
     #################################################
 
-    def _self_clmatrix_c3_c4(self, pf, order, max_shift, shift_step):
-        # return sclmatrix
-        pass
+    def _self_clmatrix_c3_c4(self):
+        """
+        Find the single pair of self-common-lines in each image assuming that the underlying
+        symmetry is C3 or C4.
+        """
+        pf = self.pf.copy()
+        n_img = self.n_img
+        n_theta = self.n_theta
+        max_shift_1d = np.ceil(2 * np.sqrt(2) * self.max_shift)
+        shift_step = self.shift_step
+        order = self.order
 
-    def _estimate_all_Riis_c3_c4(order, sclmatrix, n_theta):
-        # return Riis
-        pass
+        # The angle between self-common-lines is in the range [60, 180] for C3 symmetry
```

Comment:
This follows from Lemma A.2 in the paper. The quick answer is that a pair of self common lines arise from comparing a projection of a molecule with the projection of that molecule rotated by its symmetry. Thus the angle between two self common lines is constrained by that symmetry. 

We use 175 because any two lines that are 180 apart will be perfectly correlated.

Reply:
> This follows from Lemma A.2 in the paper. The quick answer is that a pair of self common lines arise from comparing a projection of a molecule with the projection of that molecule rotated by its symmetry. Thus the angle between two self common lines is constrained by that symmetry.

Ok. Would be good to include that as a comment.

> We use 175 because any two lines that are 180 apart will be perfectly correlated.

Ok but why 175? Wouldn't 179 or 179.9999 work better? Also the range above should then be [60, 180) (open on the right endpoint).

Reply:
> Ok. Would be good to include that as a comment.

Added a comment with brief explanation and reference to the Lemma.

> Ok but why 175? Wouldn't 179 or 179.9999 work better? Also the range above should then be [60, 180) (open on the right endpoint).

This is good question that I should probably pose to Gabi. I've made a note in the comments that these ranges were [60, 165] for C3 and [90, 160] for C4 in the MATLAB code. 

My thinking was that, due to discretization, if you get too close to 180 you chance correlating with an antipodal line.

Reply:
Is there some way of explicitly excluding the antipodal line during checking? These approaches here seem a bit kludgy.

Reply:
Ok, after some experimenting with antipodal lines of an image of just random noise I've convinced myself that a reasonable upper bound here is `180 - np.ceil(max(360/L, 360/n_theta))`. We can discuss in the dev meeting.